File: veriblock-core/src/main/java/org/veriblock/sdk/models/Coin.java
// VeriBlock Blockchain Project
// Copyright 2017-2018 VeriBlock, Inc
// Copyright 2018-2019 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package org.veriblock.sdk.models;

import org.veriblock.sdk.util.StreamUtils;
import org.veriblock.sdk.util.Utils;

import java.nio.ByteBuffer;

public final strictfp class Coin implements Comparable<Coin> {
    public static final long COIN_VALUE = 100000000;

    public static final Coin ZERO = Coin.valueOf(0);
    public static final Coin ONE = Coin.valueOf(COIN_VALUE);

    private final long atomicUnits;
    public long getAtomicUnits() {
        return atomicUnits;
    }

    private Coin(long atomicUnits) {
        this.atomicUnits = atomicUnits;
    }

    public Coin add(Coin addend) {
        return new Coin(atomicUnits + addend.atomicUnits);
    }

    public Coin subtract(Coin subtrahend) {
        return new Coin(atomicUnits - subtrahend.atomicUnits);
    }

    public Coin negate() {
        return Coin.ZERO.subtract(this);
    }

    public double toDecimal() {
        return (double) atomicUnits / (double) COIN_VALUE;
    }

    public static Coin valueOf(long atomicUnits) {
        return new Coin(atomicUnits);
    }

    @Override
    public int hashCode() {
        return (int) this.atomicUnits;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || !(obj instanceof Coin)) return false;

        return this.atomicUnits == ((Coin) obj).atomicUnits;
    }

    @Override
    public String toString() {
        return Long.toString(this.atomicUnits);
    }

    @Override
    public int compareTo(Coin o) {
        return Long.compare(atomicUnits, o.atomicUnits);
    }

    public static Coin parse(ByteBuffer txBuffer) {
        long atomicUnits = Utils.toLong(StreamUtils.getSingleByteLengthValue(txBuffer, 0, 8));
        return Coin.valueOf(atomicUnits);
    }
}

File: veriblock-core/src/main/java/org/veriblock/sdk/models/Coin.kt
// VeriBlock Blockchain Project
// Copyright 2017-2018 VeriBlock, Inc
// Copyright 2018-2019 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package org.veriblock.sdk.models

import org.veriblock.core.utilities.Utility
import org.veriblock.sdk.util.getSingleByteLengthValue
import org.veriblock.sdk.util.writeSingleByteLengthValue
import java.io.ByteArrayOutputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import kotlin.math.roundToLong

class Coin(
    val atomicUnits: Long
) : Comparable<Coin> {

    @Strictfp
    fun add(addend: Coin): Coin {
        return Coin(atomicUnits + addend.atomicUnits)
    }

    @Strictfp
    fun subtract(subtrahend: Coin): Coin {
        return Coin(atomicUnits - subtrahend.atomicUnits)
    }

    @Strictfp
    fun negate(): Coin {
        return ZERO.subtract(this)
    }

    @Strictfp
    fun toDecimal(): Double {
        return atomicUnits.toDouble() / COIN_VALUE.toDouble()
    }

    override fun hashCode(): Int {
        return atomicUnits.toInt()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        return if (other == null || other !is Coin) false else atomicUnits == other.atomicUnits
    }

    override fun toString(): String {
        return Utility.formatAtomicLongWithDecimal(atomicUnits)
    }

    override fun compareTo(other: Coin): Int {
        return atomicUnits.compareTo(other.atomicUnits)
    }

    companion object {
        const val COIN_VALUE: Long = 100000000L
        val ZERO = 0.asCoin()
        val ONE = COIN_VALUE.asCoin()

        @JvmStatic
        fun parse(buffer: ByteBuffer): Coin {
            return buffer.parseCoin()
        }
    }

    fun serialize(): ByteArray {
        ByteArrayOutputStream().use { stream ->
            serialize(stream)
            return stream.toByteArray()
        }
    }

    fun serialize(stream: OutputStream) {
        stream.writeSingleByteLengthValue(atomicUnits)
    }
}

operator fun Coin.plus(other: Coin): Coin = add(other)
operator fun Coin.minus(other: Coin): Coin = subtract(other)
operator fun Coin.unaryMinus(): Coin = negate()

fun Coin.abs(): Coin = if (this > Coin.ZERO) {
    this
} else {
    negate()
}

fun Coin.formatted(): String = toString()
fun Int.asCoin(): Coin = Coin(this.toLong())
fun Long.asCoin(): Coin = Coin(this)

fun Int.asCoinDecimal(): Coin = Coin(this * Coin.COIN_VALUE)
fun Long.asCoinDecimal(): Coin = Coin(this * Coin.COIN_VALUE)
fun Double.asCoinDecimal(): Coin = Coin((this * Coin.COIN_VALUE).roundToLong())

fun ByteBuffer.parseCoin(): Coin {
    val atomicUnits = Utility.toLong(getSingleByteLengthValue(0, 8))
    return atomicUnits.asCoin()
}
